katdal/test/test_chunkstore_s3.py

```diff
@@ -175,34 +175,75 @@ def test_jwt_broken_token(self):
+            decode_jwt(token.replace('.', ''))
         # Token header failed to decode
-        assert_raises(InvalidToken, decode_jwt, token[1:])
+        with pytest.raises(InvalidToken, match=r'Could not decode'):
+            decode_jwt(token[1:])
         # Token payload failed to decode
         h, p, s = token.split('.')
-        assert_raises(InvalidToken, decode_jwt, '.'.join((h, p[:-1], s)))
+        with pytest.raises(InvalidToken, match=r'Could not decode'):
+            decode_jwt('.'.join((h, p[:-1], s)))
         # Token signature failed to decode or wrong length
-        assert_raises(InvalidToken, decode_jwt, token[:-1])
-        assert_raises(InvalidToken, decode_jwt, token[:-2])
-        assert_raises(InvalidToken, decode_jwt, token + token[-4:])
+        with pytest.raises(InvalidToken, match=r'too short'):
+            decode_jwt(token[:-1])
+        with pytest.raises(InvalidToken, match=r'too short'):
+            decode_jwt(token[:-2])
+        with pytest.raises(InvalidToken, match=r'too long'):
+            decode_jwt(token + token[-4:])
 
     def test_jwt_expired_token(self):
         payload = {'exp': 0, 'iss': 'kat', 'prefix': ['123']}
-        token = encode_jwt(payload)
-        assert_raises(InvalidToken, decode_jwt, token)
+        with pytest.raises(InvalidToken, match=r'Token expired at 01-Jan-1970'):
+            decode_jwt(encode_jwt(payload))
         # Check that expiration time is not-too-large integer
-        payload['exp'] = 1.2
-        assert_raises(InvalidToken, decode_jwt, encode_jwt(payload))
+        payload['exp'] = '1.2'
+        with pytest.raises(InvalidToken, match=r'Expiration time must be an integer'):
+            decode_jwt(encode_jwt(payload))
         payload['exp'] = 12345678901234567890
-        assert_raises(InvalidToken, decode_jwt, encode_jwt(payload))
+        with pytest.raises(InvalidToken, match=r'an integer that is not too large'):
+            decode_jwt(encode_jwt(payload))
         # Check that it works without expiry date too
         del payload['exp']
         claims = decode_jwt(encode_jwt(payload))
-        assert_equal(payload, claims)
+        assert payload == claims
+
+
+class UnexpectedDuration(AssertionError):
+    """Test took either too short or too long to complete."""
+
+
+def duration(minimum, maximum=None):
+    """Test must finish within `minimum` to `maximum` seconds to pass.
+
+    Example use::
+
+      @duration(.2, .3)
+      def test_that_fails():
+          time.sleep(.4)
+
+    The default `maximum` differs from `minimum` by a small tolerance.
+    The check only happens if the module-level variable `check_durations` is True.
+    """
+    if maximum is None:
+        maximum = minimum + TEST_DURATION_TOLERANCE
+
+    def decorate(func):
+        def newfunc(*arg, **kw):
+            start = time.time()
+            result = func(*arg, **kw)
+            duration = time.time() - start
```

Comment:
`time.monotonic` is the recommended way to measure short elapsed time.

Reply:
Thanks for chipping in, Bruce, much appreciated! Ah, the old leap seconds :-)

Reply:
(now obsolete)